fastplotlib/graphics/_vectors.py

```diff
@@ -0,0 +1,391 @@
+            if set(vector_shape_options.keys()) != required:
+                raise KeyError(
+                    f"`vector_shape_options` must be a dict with the following keys: {required}.\n"
+                    f"You have passed: {vector_shape_options}"
+                )
+            shape_options = vector_shape_options
+        else:
+            if size is None:
+                # guess from density
+                # sort xs and then take unique to get the density along x, same for y and z
+                x_density = np.diff(np.unique(np.sort(self._positions[:, 0]))).mean()
+                y_density = np.diff(np.unique(np.sort(self._positions[:, 1]))).mean()
+                densities = [x_density, y_density]
+
+                # if z is not basically zero
+                if not np.allclose(
+                    np.diff(np.unique(np.sort(self._positions[:, 2]))), 0.0
+                ):
+                    z_density = np.diff(np.unique(np.sort(positions[:, 2]))).mean()
+                    densities.append(z_density)
+
+                mean_density = np.mean(densities)
+
+                size = mean_density
+
+            cone_height = size / 2
+            stalk_height = size / 2
+
+            cone_radius = size / 10
+            stalk_radius = cone_radius / 8
+
+            shape_options = {
+                "cone_radius": cone_radius,
+                "cone_height": cone_height,
+                "stalk_radius": stalk_radius,
+                "stalk_height": stalk_height,
+            }
+
+        geometry = create_vector_geometry(color=color, **shape_options)
+        material = pygfx.MeshBasicMaterial()
+
+        n_vectors = self._positions.value.shape[0]
+
+        world_object = pygfx.InstancedMesh(geometry, material, n_vectors)
+
+        magnitudes = np.linalg.norm(self.directions[:], axis=1, ord=2)
+
+        for i in range(n_vectors):
+            # get quaternion to rotate vector to new direction
+            rotation = la.quat_from_vecs(
+                self._directions.init_direction, self._directions[i]
+            )
+            # get the new transform
+            transform = la.mat_compose(
+                self._positions.value[i], rotation, magnitudes[i]
+            )
+            # set the buffer
+            world_object.instance_buffer.data["matrix"][i] = transform.T
+
+        world_object.instance_buffer.update_full()
+
+        self._set_world_object(world_object)
+
+    @property
+    def positions(self) -> VectorPositions:
+        """Vector positions"""
+        return self._positions
+
+    @positions.setter
+    def positions(self, new_positions):
+        self._positions.set_value(self, new_positions)
+
+    @property
+    def directions(self) -> VectorDirections:
+        """Vector directions"""
+        return self._directions
+
+    @directions.setter
+    def directions(self, new_directions):
+        self._directions.set_value(self, new_directions)
+
+
```

Comment:
Can you add a docstring to the three functions below? Even if they are mostly just for internal use, it would be helpful for me to better understand what each of these is doing.

Reply:
added a few